[XylusChen] iP

XylusChen opened the pull request:
## XyDuke
> "Too much work, too little time. Xyduke's here and you'll be fine!" - Xylus Chen

Xyduke saves you so much time by keeping track of all your task and commitments! It's:
* straightforward
* fool-proof
* ~~FAST~~ **INSANELY FAST** to use

Let me help you get started. Here's what you need to do:
1. Download it from [here](https://github.com/XylusChen/ip/releases/tag/A-Jar).
2. Double-Click the jar file. Make sure you have **JAVA** installed locally.
3. Explore XyDuke's features.
4. Enjoy the benefits!

Most importantly... IT IS ***FREEEEEEE!*** :star_struck::heart_eyes::smiling_face_with_three_hearts:

Here are the key features of XyDuke:
- [x] Managing Todo, Deadline and Event tasks.
- [x] Marking tasks as complete or incomplete.
- [x] Deleting tasks.
- [x] Searching tasks using keywords.
- [ ] Sorting tasks by Date. (*coming soon*)

If you are a fellow Duke developer, do let me know how you feel about XyDuke! Here is my `main` method:
```
public static void main(String[] args) {
    new Duke().run();
}
```

Commit:
Add Gradle support

XylusChen committed:
changed greeting message

XylusChen committed:
XyDuke now echos user commands, exit when "bye"

XylusChen committed:
stores text entered by user, lists tasks when requested

XylusChen committed:
added task class, can now mark or unmark tasks as done

XylusChen committed:
added support for multiple task types using inheritance

XylusChen committed:
i/o redirection technique to semi-automate testing of Duke

XylusChen committed:
handle errors using exceptions: invalid commands, incomplete commands

XylusChen committed:
added support for deleting tasks, fixed bug where task descriptions cannot contain commands

XylusChen committed:
updated automated testing to test delete and error handling

XylusChen committed:
fixed bug where inputs of the form <command> + <string> where string does not begin with a whitespace is not recognized as invalid

XylusChen committed:
add tasktypes package

XylusChen committed:
Task class: Added numTasks static variable.

Use static variable Task.numTasks to keep track of number of tasks,
instead of checking task's array size.

Good coding practice, follows OOP principles better.

XylusChen committed:
Add file saving capabilities.

Tasks are saved in duke.txt. Tasks are loaded from hard disk when
Duke boots up.

XylusChen committed:
Add Dates and Times support

Duke can now parse date and time String inputs into LocalDateTime
objects.

XylusChen committed:
Merge branch 'branch-Level-7'

XylusChen committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/tasktypes/Deadline.java
#	src/main/java/tasktypes/Event.java

XylusChen committed:
Bug fix for delete command

Task number from user input should not be more than number of tasks
available.

XylusChen committed:
Add TaskList class

Extracted TaskList class from Duke class. TaskList object handles
all task list related operations.

XylusChen committed:
Add Storage class

Extracted Storage class from Duke class. Storage class handles all
loading and saving tasks to data file.

XylusChen committed:
Added Ui class

Extracted Ui class from Duke class. Ui class handles all
interactions with user.

XylusChen committed:
Add Parser and Command classes

Parser class makes sense of user input (command).
Command object represents a type of Command, each having unique
execute behavior.

XylusChen committed:
Put all classes into package duke. Further divide into multiple
packages.

XylusChen committed:
Merge remote-tracking branch 'origin/add-gradle-support'

XylusChen committed:
Add Gradle support

Integrate Gradle into IntelliJ. Build and Run Duke using gradle.

XylusChen committed:
Add JUnit tests to test the behavior of code.

Implement test for ToDo and Deadline Tasks, check correctness
of String conversion and Save Format conversion.

XylusChen committed:
build.gradle: Change mainClassName variable to "duke.Duke"

XylusChen committed:
Privatize fields to obey information hiding. Change Task Class to
abstract class.

Fields in Command classes and TaskTypes classes are set to private.

XylusChen committed:
Add JavaDocs

XylusChen committed:
Tweak code to comply with given coding standard.

XylusChen committed:
Change indentation for Switch Case.

XylusChen committed:
Add Find Command support.

Users can now search for matching tasks using keywords or phrases.

XylusChen committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/parser/Parser.java

XylusChen committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/tasklist/TaskList.java

XylusChen committed:
Update JavaDocs

XylusChen committed:
Update .gitignore to ignore all /text-ui-test/ files

XylusChen committed:
Update .gitignore

XylusChen committed:
Update .gitignore

XylusChen committed:
Update .gitignore

XylusChen committed:
Add CheckStyle support for Duke.

Include Gradle checkstyle script and CheckStyle-IDEA plugin.

XylusChen committed:
FindCommand: Add VarArgs support for initializing FindCommand
instance.

Support searching for multiple keywords.

XylusChen committed:
Merge branch 'branch-A-CheckStyle'

XylusChen committed:
Fix bug for Deadline and Event tasks where invalid date/time input
will result in Duke crashing.

XylusChen committed:
Add GUI support for Duke.

XylusChen committed:
Merge branch 'branch-Level-10'

XylusChen committed:
Update JavaDocs for methods changed to support GUI implementation.

Jnjy commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+ * Represents a delete task command.
+ */
+public class DeleteCommand extends Command {
+    private int toDelete;
```
Perhaps more intuitive variable name can be used here? 

My suggestion would be `indexToDelete`

I noticed the same issue in `MarkCommand` and `UnmarkCommand` as well.

XylusChen replied:
You're right. I will fix that, thanks!

Jnjy commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,178 @@
+        sc = new Scanner(System.in);
+    }
+
+    /**
+     * Closes the Scanner object
+     */
+    public void close() {
+        sc.close();
+    }
```
Any reason why you have another `close()` function instead of closing the Scanner object in `goodbye()`?

XylusChen replied:
I try to use the UI instance as much as possible, instead of interacting with the Scanner object directly. Hence the wrapping of ```sc.close()``` in a ```ui.close()``` method.

Jnjy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public void run() {
+
+        ui.showLine();
+        ui.greet();
+        ui.showLine();
+
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String fullCommand = ui.nextInput();
+                ui.showLine();
+                Command c = Parser.parse(fullCommand, tasks);
+                c.execute(tasks, storage, ui);
+                isExit = c.isGoodbye();
+            } catch (DukeException e) {
+                ui.showError(e);
+            } finally {
+                ui.showLine();
+            }
+        }
```
I like how you separate input reading and output printing into another `UI` class which makes your code cleaner!

XylusChen replied:
Thank you!

Jnjy commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class ExitCommand extends Command {
+
+    public ExitCommand() {
+    }
```
Perhaps you don't need to include this as it is just a default constructor.

XylusChen replied:
You're right! I will fix that!

Jnjy commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,155 @@
+            inputArr = fullCommand.split(" ");
+            int toUnmark = Integer.parseInt(inputArr[1]);
+            if (toUnmark > tasks.getNumTasks()) {
+                throw new DukeException("Task does not exist! Please enter valid input!");
```
Perhaps you can extract the different `Exceptions` into different classes? 

Suggestion would be:
```
DukeException (Base Class)
|_ InvalidArgumentException
|_ MissingArgumentException
|_ InvalidDateException
|_ (etc)
```

XylusChen replied:
Good suggestion. I will consider doing that!

Jnjy commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,43 @@
+import duke.tasklist.TaskList;
+import duke.tasktypes.Task;
+import duke.ui.Ui;
+
+
```
Minor comment: Extra breakline is inserted

XylusChen replied:
Thanks for highlighting. Will fix!

Jnjy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,142 @@
+ */
+public class Storage {
+
+    private static final String DEFAULT_DIRECTORY = "./data/";
+    private static final String DEFAULT_FILEPATH = "./data/duke.txt";
```
I like how your code adhere to coding standard!

XylusChen replied:
Thank you!

Jnjy reviewed (commented):
Overall, great work! I like how your overall code looks clean and extractions are done properly for ease of code reading. Keep up the good work!

XylusChen committed:
Merge remote-tracking branch 'origin/master'

ScorpiusSigma commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,155 @@
+ */
+public class Parser {
+
+    private static final String[] VALUE_COMMANDS = {"unmark ", "mark ", "delete ", "find "};
+    private static final String[] TASK_COMMANDS = {"deadline ", "todo ", "event "};
```
Perhaps you can use an ENUM rather than an array for the commands representation?

Maybe something like this:
```java
enum ValueCommands {
  UNMARK,
  MARK,
  DELETE,
  FIND
}
```

ScorpiusSigma commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @param toDelete Integer index of task to be deleted.
+     * @return Task has been deleted message.
+     */
```
I love how you did your documentations, it super clean! 🧼

ScorpiusSigma commented on src/main/java/duke/tasktypes/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+    public Deadline(String description, String by) throws DukeException {
+        super(description);
+        String[] dateAndTime = by.split(" ");
+        String date = dateAndTime[0];
+        String time = dateAndTime[1];
+        date = date.replace('/', '-');
+        this.forSaving = date + " " + time;
+        time = time.substring(0, 2) + ':' + time.substring(2);
```
Interesting way of getting manipulating the string to get your time. Would it be better if you have a DatetimeParser class that would handle the string input and return you time ?

ScorpiusSigma commented on src/main/java/duke/tasktypes/Event.java:
```diff
@@ -0,0 +1,93 @@
+                        .format(this.startDateTime);
+
+
+        String[] endDateAndTime = end.split(" ");
+        String endDate = endDateAndTime[0];
+        String endTime = endDateAndTime[1];
+        endDate = endDate.replace('/', '-');
+        this.forSavingEnd = endDate + " " + endTime;
+        endTime = endTime.substring(0, 2) + ':' + endTime.substring(2);
```
I have given my opinion above on how you can handle string to datetime string to make your method shorter. I hope it will be useful for you

ScorpiusSigma reviewed (commented):
I love your overall code base on how clean and well documented it is! I have included recommended on how to handle your datetime and I hope it helps :)

lukkesreysandeur commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+
+        ui.showLine();
+        ui.greet();
+        ui.showLine();
```
I like how the code is formatted logically with separations in different levels of abstraction.

lukkesreysandeur commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+            return output;
+        } catch (DukeException de) {
+            return ui.showError(de);
+        } catch (Exception e) {
```
May I ask what this catch statement is for?

lukkesreysandeur commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+        this.toDelete = toDelete;
+    }
+
+    @Override
```
Perhaps instead of leaving the header comment you might want to use the inheritDoc annotation? Same for all the other Command classes.

lukkesreysandeur commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public String execute(TaskList tasks, Storage storage, Ui ui) {
+        TaskList matchTasks = tasks.getMatchingTasks(this.keywords);
```
I think that the this can be removed as it is the only keyword in this scope.

lukkesreysandeur commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class MarkCommand extends Command {
+
+    private int toMark;
```
Maybe a more intuitive name would be something like index or indexToMark?

lukkesreysandeur commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,155 @@
+    private static final String[] VALUE_COMMANDS = {"unmark ", "mark ", "delete ", "find "};
+    private static final String[] TASK_COMMANDS = {"deadline ", "todo ", "event "};
+
+    public Parser() {
```
Perhaps consider removing the empty constructor if you are not using it.

lukkesreysandeur commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,155 @@
+            }
+            return new UnmarkCommand(toUnmark);
+        case "mark ":
+            inputArr = fullCommand.split(" ");
```
Perhaps consider abstracting the handling of user input to another method so you won't need to repeat this code in every switch case.

lukkesreysandeur commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,142 @@
+ */
+public class Storage {
+
+    private static final String DEFAULT_DIRECTORY = "./data/";
```
Good use of constants, I might do that for my code as well.

lukkesreysandeur commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+ * A TaskList object corresponds to User's task collection represented by an ArrayList.
+ */
+public class TaskList {
+    private ArrayList<Task> taskStorage;
```
Might actually be more intuitive to just name this as tasks?

lukkesreysandeur commented on src/main/java/duke/tasktypes/Event.java:
```diff
@@ -0,0 +1,93 @@
+
+    private LocalDate startDate;
+    private LocalTime startTime;
+    private LocalDateTime startDateTime;
```
Maybe consider saving the date and time in a single LocalDateTime object?

lukkesreysandeur commented on src/main/java/duke/tasktypes/Event.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public String getSaveFormat() {
+        String done;
+        if (this.isDone) {
+            done = "1";
+        } else {
+            done = "0";
+        }
```
You could consider maybe abstracting this to a single statement with the ternary operator. Something like
```suggestion
        String done = isDone ? "1" : "0";
```
Should make the code more readable.

lukkesreysandeur commented on src/main/java/duke/tasktypes/Task.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Mark a task as incomplete.
+     */
+    public void setUndone() {
```
Small nit but maybe you can consider renaming to setNotDone?

lukkesreysandeur commented on src/main/java/duke/tasktypes/ToDo.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Represents a Todo Task.
+ */
+public class ToDo extends Task {
```
Maybe you should rename the class to simply Todo.

lukkesreysandeur reviewed (commented):
Overall I like the quality of the code. It is mostly intuitive and easily read. I think you could consider adding more exceptions classes but other than that I don't see many issues.

XylusChen committed:
Add Java Assertions to XyDuke

There are several methods across Parser, Storage and TaskList
classes that could user an assert statement.

Adding assert statements improves the method's reliability and
the reliability of XyDuke as a whole.

Let's identify areas in code where assert statements could benefit
by ensuring program correctness at that point and prevent further
build up of incorrect program state.

XylusChen committed:
Improve on code quality through refactoring.

There are several methods that are excessively long and untidy. This
severely affects code readability and should be improved.

Transforming long methods into a number of small methods improves
code quality.

As a step towards further refactoring, let's extract functionalities
into separate short methods in their respective classes. Doing so
shortens lengthy methods and makes code more readable.

XylusChen committed:
Merge pull request #1 from XylusChen/branch-A-CodeQuality

Improve on code quality through refactoring.

XylusChen committed:
Merge branch 'branch-A-CodeQuality'

XylusChen committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/tasklist/TaskList.java

XylusChen committed:
Merge pull request #2 from XylusChen/branch-A-Assertions

Add Java Assertions to XyDuke

XylusChen committed:
Merge branch 'branch-A-Assertions'

XylusChen committed:
Set up Continuous Integration using GitHub Actions

XylusChen committed:
Add "orderedlist" command

Print tasks ordered by deadline.

XylusChen committed:
Fix CheckStyle errors

XylusChen committed:
Add method to reset task ID counter variable

Implemented to ensure consistency in test cases for String conversion.

XylusChen committed:
Merge branch 'master' of https://github.com/XylusChen/ip

* 'master' of https://github.com/XylusChen/ip:
  Add method to reset task ID counter variable
  Fix CheckStyle errors
  Add "orderedlist" command
  Set up Continuous Integration using GitHub Actions
  Improve on code quality through refactoring.
  Add Java Assertions to XyDuke

XylusChen committed:
Add changes to GUI and Duke's personality.

Custom images and icons for Duke's GUI.
Crop circular images for DialogBox.
Renamed Duke to XyDuke.
Change font color to discern User and XyDuke's messages.
Change font size to make text more visible.
Change ScrollPane background color to make GUI more pleasing.

XylusChen committed:
Add colour to User Input TextField.

XylusChen committed:
Add error handler for invalid deadline or event inputs

Improve error handling for adding deadline or event tasks that are
missing a DateTime input.

XylusChen committed:
Add close feature

GUI logic now detects ExitCommand. Closes GUI window upon saving
current state of tasks.

XylusChen committed:
Add XyDuke product screenshot

XylusChen committed:
Update README.md

Add User Guide for XyDuke.

XylusChen committed:
Rename Ui.png

XylusChen committed:
Merge branch 'master' of https://github.com/XylusChen/ip

* 'master' of https://github.com/XylusChen/ip:
  Update README.md

XylusChen committed:
Add Label background and border

XylusChen committed:
XyDuke is now XyDuke

XylusChen committed:
Create _config.yml

Add emoji support for website.

XylusChen committed:
Update _config.yml

Add theme Minima

XylusChen committed:
Update _config.yml

Midnight theme

XylusChen committed:
Merge branch 'master' of https://github.com/XylusChen/ip

* 'master' of https://github.com/XylusChen/ip:
  Update _config.yml
  Update _config.yml
  Create _config.yml

XylusChen committed:
Update _config.yml

Change to midnight theme

XylusChen committed:
Merge branch 'master' of https://github.com/XylusChen/ip

* 'master' of https://github.com/XylusChen/ip:
  Update _config.yml

XylusChen committed:
Fix checkstyle errors and JUnit test errors.

XylusChen committed:
Update GitHub actions and update product screenshot

XylusChen committed:
Change to Hacker theme got GitHub Pages

XylusChen committed:
Update midnight theme configuration

XylusChen committed:
Update DialogBox settings, size now changes dynamically to display
all text.

XylusChen committed:
Merge branch 'master' of https://github.com/XylusChen/ip

* 'master' of https://github.com/XylusChen/ip:
  Update DialogBox settings, size now changes dynamically to display all text.
  Update midnight theme configuration
  Change to Hacker theme got GitHub Pages
  Update GitHub actions and update product screenshot
  Fix checkstyle errors and JUnit test errors.
  Update _config.yml
  Update _config.yml
  Update _config.yml
  Create _config.yml
  XyDuke is now XyDuke
  Add Label background and border
  Rename Ui.png
  Update README.md
  Add XyDuke product screenshot
  Add close feature
  Add error handler for invalid deadline or event inputs
  Add colour to User Input TextField.
  Add changes to GUI and Duke's personality.

XylusChen committed:
Update Duke storage system

Duke now saves after every modify command.

XylusChen committed:
Update input error check

Duke now checks for invalid task ID input type for
mark, unmark, delete commands.

XylusChen committed:
Fix CheckStyle issues
